### ISSUE TITLE
Upgrade to version 3.0.4 breaks gitlab

### ISSUE THREAD

**Issue by mfechner** [FooBarWidget/default_value_for#72]
Hi,

the upgrade from version 3.0.3 to 3.0.4 breaks the gitlab installation.

You can reproduce it by executing:
```
su -l git -c "cd /usr/local/www/gitlab && rake --trace gitlab:env:info RAILS_ENV=production"
** Invoke gitlab:env:info (first_time)
** Invoke environment (first_time)
** Execute environment
The PGconn, PGresult, and PGError constants are deprecated, and will be
removed as of version 1.0.

You should use PG::Connection, PG::Result, and PG::Error instead, respectively.

Called from /usr/local/lib/ruby/gems/2.4/gems/activesupport-4.2.10/lib/active_support/dependencies.rb:240:in `load_dependency'
** Execute gitlab:env:info

System information
System:
Current User:   git
Using RVM:      no
Ruby Version:   2.4.3p205
Gem Version:    2.7.4
Bundler Version:1.16.1
Rake Version:   12.3.0
Redis Version:  3.2.11
Git Version:    2.16.1
Sidekiq Version:5.1.0
Go Version:     go1.9.3 freebsd/amd64
rake aborted!
NoMethodError: undefined method `issues_enabled' for #<Project id: some-group/some-project>
/usr/local/lib/ruby/gems/2.4/gems/activemodel-4.2.10/lib/active_model/attribute_methods.rb:433:in `method_missing'
/usr/local/lib/ruby/gems/2.4/gems/default_value_for-3.0.4/lib/default_value_for.rb:163:in `block in set_default_values'
/usr/local/lib/ruby/gems/2.4/gems/default_value_for-3.0.4/lib/default_value_for.rb:154:in `each'
/usr/local/lib/ruby/gems/2.4/gems/default_value_for-3.0.4/lib/default_value_for.rb:154:in `set_default_values'
/usr/local/lib/ruby/gems/2.4/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:432:in `block in make_lambda'
/usr/local/lib/ruby/gems/2.4/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:228:in `block in halting_and_conditional'
/usr/local/lib/ruby/gems/2.4/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:506:in `block in call'
/usr/local/lib/ruby/gems/2.4/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:506:in `each'
/usr/local/lib/ruby/gems/2.4/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:506:in `call'
/usr/local/lib/ruby/gems/2.4/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:92:in `__run_callbacks__'
/usr/local/lib/ruby/gems/2.4/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:778:in `_run_initialize_callbacks'
/usr/local/lib/ruby/gems/2.4/gems/activerecord-4.2.10/lib/active_record/core.rb:284:in `initialize'
/usr/local/lib/ruby/gems/2.4/gems/default_value_for-3.0.4/lib/default_value_for.rb:142:in `initialize'
/usr/local/lib/ruby/gems/2.4/gems/state_machines-activerecord-0.5.0/lib/state_machines/integrations/active_record.rb:463:in `initialize'
/usr/local/lib/ruby/gems/2.4/gems/activerecord-4.2.10/lib/active_record/inheritance.rb:61:in `new'
/usr/local/lib/ruby/gems/2.4/gems/activerecord-4.2.10/lib/active_record/inheritance.rb:61:in `new'
/usr/local/lib/ruby/gems/2.4/gems/activerecord-4.2.10/lib/active_record/reflection.rb:141:in `build_association'
/usr/local/lib/ruby/gems/2.4/gems/activerecord-4.2.10/lib/active_record/associations/association.rb:250:in `build_record'
/usr/local/lib/ruby/gems/2.4/gems/activerecord-4.2.10/lib/active_record/associations/collection_association.rb:157:in `build'
/usr/local/lib/ruby/gems/2.4/gems/activerecord-4.2.10/lib/active_record/associations/collection_proxy.rb:259:in `build'
/usr/local/www/gitlab/lib/tasks/gitlab/info.rake:40:in `block (3 levels) in <top (required)>'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/task.rb:251:in `block in execute'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/task.rb:251:in `each'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/task.rb:251:in `execute'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/task.rb:195:in `block in invoke_with_call_chain'
/usr/local/lib/ruby/2.4/monitor.rb:214:in `mon_synchronize'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/task.rb:181:in `invoke'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:160:in `invoke_task'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:116:in `each'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:116:in `block in top_level'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:125:in `run_with_threads'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:110:in `top_level'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:83:in `block in run'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:186:in `standard_exception_handling'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/lib/rake/application.rb:80:in `run'
/usr/local/lib/ruby/gems/2.4/gems/rake-12.3.0/exe/rake:27:in `<top (required)>'
/usr/local/bin/rake:23:in `load'
/usr/local/bin/rake:23:in `<main>'
Tasks: TOP => gitlab:env:info

```
I'm not sure if this is related to gitlab, but doing a downgrade to version 3.0.3 bring gitlab back to live.

**Comment by chrisarcand** [comment]
By GitLab, I'm assuming you mean the project here: https://github.com/gitlabhq/gitlabhq

This is undoubtedly caused by #63 (the only thing included in 3.0.4). `Project` has a writer for `issues_enabled=` but no reader `issues_enabled`: https://github.com/gitlabhq/gitlabhq/blob/498ade4801a822f8704390b10d178af9fe7987cb/app/models/concerns/project_features_compatibility.rb#L22-L24

Based on [the README here](https://github.com/FooBarWidget/default_value_for#attributes-that-arent-database-columns), one shouldn't have to specify a reader for `default_value_for` to work, so I'd consider this a bug.

cc/ @njakobsen , do you have a suggestion for how to address this? I can take a quick glance but otherwise we might just have to yank 3.0.4 and revert the fix made in #63 until a better solution can be found.

**Comment by njakobsen** [comment]
I think the general behaviour pattern for ActiveRecord::Base attributes is identically named reader/writer method pairs, not direct access to the attributes hash, so I would consider calling setters and getters standard practice. On top of that sits the store columns, which, although not backed by the db, follow the same pattern of access. Having a setter without a getter with a matching name seems to break this convention, and I would argue that in that case, an additional config option would be appropriate. e.g. `default_value_for :issues_enabled, reader: -> { @issues_enabled }`. The alternative would be to add a setter to match the getter in `ProjectFeaturesCompatibility`.

As for the breaking change, @chrisarcand, maybe #63 should be a minor release instead of a patch?

**Comment by chrisarcand** [comment]
> I think the general behaviour pattern for ActiveRecord::Base attributes is identically named reader/writer method pairs, not direct access to the attributes hash, so I would consider calling setters and getters standard practice. 

I don't disagree. I think the behavior that I cite [here](https://github.com/FooBarWidget/default_value_for/tree/1d1b2592eb00741a290d8406353846ddd8aae335#attributes-that-arent-database-columns) ('this gem should work even if you have a writer method with *no* reader method') is actually really awkward and I wouldn't support it. However, it's explicitly stated and removing that behavior is a breaking change worthy of a major version bump (and I'm not against that - just stating facts), which means *if* we were to keep everything as-is in #63, that's not a minor (or a patch) release.

I've opened #73 which should address this regression while still incorporating your fix in #63 (thus, no breaking changes or added behavior). I'm running a few more tests across Rails versions locally (shame this project doesn't have CI...); then I'll cut 3.0.5 tomorrow and yank 3.0.4.

**Comment by chrisarcand** [comment]
@mfechner I've released 3.0.5 which should fix this issue. Please reopen this if it doesn't. Thanks!

**Comment by mfechner** [comment]
@chrisarcand thanks a lot for this quick fix!
I updated it already in the FreeBSD port for the new gitlab 10.4 release.
